File: community/cypher/cypher-backport/2.0/cypher-compiler-backport-2.0/src/test/scala/org/neo4j/cypher/internal/compiler/v2_0/TypeTest.scala
/**
 * Copyright (c) 2002-2015 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.compiler.v2_0

import commands.expressions._
import commands.expressions.Add
import commands.expressions.Literal
import commands.expressions.Multiply
import commands.expressions.Subtract
import pipes.QueryStateHelper
import org.junit.Test
import org.hamcrest.CoreMatchers._
import org.hamcrest.Matcher

class TypeTest {
  @Test
  def plus_int_int() {
    val op = Add(Literal(1), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Integer]))
  }

  @Test
  def plus_double_int() {
    val op = Add(Literal(1.2), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Double]))
  }

  @Test
  def minus_int_int() {
    val op = Subtract(Literal(1), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Integer]))
  }

  @Test
  def minus_double_int() {
    val op = Subtract(Literal(1.2), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Double]))
  }

  @Test
  def multiply_int_int() {
    val op = Multiply(Literal(1), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Integer]))
  }

  @Test
  def multiply_double_int() {
    val op = Multiply(Literal(1.2), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Double]))
  }

  @Test
  def divide_int_int() {
    val op = Divide(Literal(1), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Integer]))
  }

  @Test
  def divide_double_int() {
    val op = Divide(Literal(1.2), Literal(2))

    val result = calc(op)

    assertThat(result, instanceOf(classOf[java.lang.Double]))
  }

  private def assertThat(x:Any, matcher:Matcher[AnyRef]) {
    org.junit.Assert.assertThat("\nGot a: " + x.getClass, x.asInstanceOf[Object], matcher)
  }

  private def calc(e:Expression) = e.apply(ExecutionContext.empty)(QueryStateHelper.empty)
}

File: community/capabilities/src/test/java/org/neo4j/capabilities/TypeTest.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.capabilities;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.neo4j.capabilities.Type.BOOLEAN;
import static org.neo4j.capabilities.Type.DOUBLE;
import static org.neo4j.capabilities.Type.FLOAT;
import static org.neo4j.capabilities.Type.INTEGER;
import static org.neo4j.capabilities.Type.LONG;
import static org.neo4j.capabilities.Type.STRING;
import static org.neo4j.capabilities.Type.listOf;

import java.util.Collection;
import org.junit.jupiter.api.Test;

class TypeTest {

    @Test
    void testString() {
        assertThat(STRING.name()).isEqualTo("string");
        assertThat(STRING.description()).isEqualTo("a string value");
        assertThat(STRING.type()).isEqualTo(String.class);
    }

    @Test
    void testBoolean() {
        assertThat(BOOLEAN.name()).isEqualTo("boolean");
        assertThat(BOOLEAN.description()).isEqualTo("a boolean value");
        assertThat(BOOLEAN.type()).isEqualTo(Boolean.class);
    }

    @Test
    void testInteger() {
        assertThat(INTEGER.name()).isEqualTo("integer");
        assertThat(INTEGER.description()).isEqualTo("an integer value");
        assertThat(INTEGER.type()).isEqualTo(Integer.class);
    }

    @Test
    void testLong() {
        assertThat(LONG.name()).isEqualTo("long");
        assertThat(LONG.description()).isEqualTo("a long value");
        assertThat(LONG.type()).isEqualTo(Long.class);
    }

    @Test
    void testFloat() {
        assertThat(FLOAT.name()).isEqualTo("float");
        assertThat(FLOAT.description()).isEqualTo("a float value");
        assertThat(FLOAT.type()).isEqualTo(Float.class);
    }

    @Test
    void testDouble() {
        assertThat(DOUBLE.name()).isEqualTo("double");
        assertThat(DOUBLE.description()).isEqualTo("a double value");
        assertThat(DOUBLE.type()).isEqualTo(Double.class);
    }

    @Test
    void testListOf() {
        var list = listOf(STRING);
        assertThat(list.name()).isEqualTo("list of string");
        assertThat(list.description()).isEqualTo("a list of string values");
        assertThat(list.type()).isEqualTo(Collection.class);

        assertThatThrownBy(() -> listOf(listOf(STRING)))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("nested list types is not supported.");
    }
}
